### PR TITLE
Move persistent store to app group

### PR DESCRIPTION
**Pull request by diegoreymendez**
Moves the persisten store to the App Group so that it can be used by libraries to implement services.

## Testing:

- Please, test all scenarios with the feature flag `.swiftCoreData` ON and OFF.  We need to make sure nothing is broken.
- For all tests I recommend to filter Xcode’s console log by this text “org.wordpress.StoreToAppGroupMigration” 

### Test 1: Fresh install

1. Delete WP from your phone
2. Run the App

Make sure you see this message in the log:

> org.wordpress.StoreToAppGroupMigration: Persistent store does not exist at source location file:///var/mobile/Containers/Data/Application/425CE222-4755-4133-8AE4-F4BBF6662C76/Documents/WordPress.sqlite. We assume this is the first App run.

3. Launch the app again.

Make sure you see this message in the log:

> org.wordpress.StoreToAppGroupMigration: Migration already performed.

### Test 2: Existing install

Do a clean run of the App by deleting it from the device.

1. Check out branch `release/16.0`.  This includes a version of the data model several versions old.
2. Run the app, and login.
3. Kill the App from the device.
4. Check out branch `try/move-persistent-store-to-app-group`.
5. Run the app.

Make sure you see this message in the log.

> org.wordpress.StoreToAppGroupMigration: Persistent store moved successfully to file:///private/var/mobile/Containers/Shared/AppGroup/93E1CD1A-0140-4B75-AC6B-EF138B346F89/WordPress.sqlite

6. Launch the app again.

Make sure you see this message in the log:

> org.wordpress.StoreToAppGroupMigration: Migration already performed.

## PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by diegoreymendez** [comment]
@Gio2018 - Before merging this we should probably implement our Widgets service to see if these changes can really be of help, and to avoid having to roll this back in the future.

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/CoreDataFileLocationManager.swift]: nit: both these imports can probably be omitted.

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/CoreDataFileLocationManager.swift:8]: nit: is there a reason why we need `baseFileName` or can we collapse into one constant?

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/CoreDataFileLocationManager.swift:34]: Can you explain the syntax and why this is different form directly calling `moveStoreFromMainBundleToAppGroup`, given that the method checks for `StoreToAppGroupMigrationCompleteKey` before the execution?

File: WordPress/Classes/Utility/CoreDataManager.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/CoreDataManager.swift]: any reason why this changed from `let` to `var`?

File: WordPress/Classes/Utility/CoreDataStoreMover.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/CoreDataStoreMover.swift]: It seems that on success, an url is returned, not a bool. Maybe the comment needs to be updated?

File: WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift]: not sure if we want to catch the error here too?

File: WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift]: Is this method for future use? Looks like it's not called anywhere.

**Review by Gio2018 — COMMENTED** [review]
Hey @diegoreymendez , thanks so much for working on this!! I still have to do the test scenarios; left some code comments. Other than those, I wonder if we should add some Sentry logging into this as well, what do you think?

File: WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift]: Nopes.  This was actually part of an approach I was taking to resolve an issue in my PR, but I resolved it another way, so I should remove this code.  Good catch.

File: WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/Migrator/CoreDataIterativeMigrator.swift]: I'll remove the entire method for now.

File: WordPress/Classes/Utility/CoreDataStoreMover.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/CoreDataStoreMover.swift]: Updated.

File: WordPress/Classes/Utility/CoreDataManager.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/CoreDataManager.swift]: Just something in need of a cleanup.  Good catch.

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/CoreDataFileLocationManager.swift]: Removed.

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/CoreDataFileLocationManager.swift:8]: I've modified line 44 to make use of this (that was the original intention).  Check it out.

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/CoreDataFileLocationManager.swift:34]: I agree this is ugly and not very intuitive.  This code I added is the equivalent of the old `dispatch_once` calls (which IMHO were easier to read).

From https://swift.org/migration-guide-swift3/:

> The free function dispatch_once is no longer available in Swift. In Swift, you can use lazily initialized globals or static properties and get the same thread-safety and called-once guarantees as dispatch_once provided. Example:
>
> `let myGlobal = { … global contains initialization in a call to a closure … }()`
> `_ = myGlobal  // using myGlobal will invoke the initialization code only `

While this is ugly, I'd still prefer to stick with this official suggestion (to replace `dispatch_once`) since it ensures thread-safety and called-once guarantees (as per the description).

**Comment by diegoreymendez** [comment]
> I wonder if we should add some Sentry logging into this as well, what do you think?

Probably, although I'm not extremely familiar with what kinds of logs we send to Sentry.  Could you ping me directly when you're online to discuss this?

Thanks for the review!  I've addressed all current feedback so I'm re-requesting the review.

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (Gio2018) [reply WordPress/Classes/Utility/CoreDataFileLocationManager.swift:34]: I see it now and makes total sense, thanks for clarifying!!

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (Gio2018) [inline WordPress/Classes/Utility/CoreDataFileLocationManager.swift]: nit: since this condition also occurs at the subsequent runs after a fresh install, should we change the message into something like: "Migration not required"?. In that case, there was no migration previously performed.

**Review by Gio2018 — APPROVED** [review]
Performed all test cases with `swiftCoreData` feature flag both ON and OFF, and all worked as expected. Left another very small comment but other than that, LGTM!

File: WordPress/Classes/Utility/CoreDataFileLocationManager.swift
Comment (diegoreymendez) [reply WordPress/Classes/Utility/CoreDataFileLocationManager.swift]: That's a good point.

**Comment by diegoreymendez** [comment]
Thanks for the review!

I'm holding off on merging this for now until we're sure it's what we need.

I'm now working to implement a service making use of the new location for the store.  Once it's up we can review it makes sense, and merge this as well.

**Comment by diegoreymendez** [comment]
After chatting with @jkmassel and @Gio2018 about this... I'm closing the PR since there are concerns about the direction taken here.  There could be concurrency issues, and it's overall not clear if this is the path we want to take.

Thanks to both for your feedback.

**Pull request closed without merging**
